subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/EdgeState.java

```diff
@@ -274,7 +280,7 @@ private static String formatSelector(SelectorState state, Attribute<?> attribute
         DependencyMetadata dependencyMetadata = state.getDependencyMetadata();
         StringBuilder sb = new StringBuilder(dependencyMetadata.isPending() ? "Constraint '" : "Dependency '");
         sb.append(state.getRequested()).append("' wants value ");
-        sb.append("'").append(dependencyMetadata.getAttributes().getAttribute(attribute)).append("'");
+        sb.append("'").append(((ModuleComponentSelector)dependencyMetadata.getSelector()).getAttributes().getAttribute(attribute)).append("'");
```

Comment:
Seems strange that the change above does an instance check but here it is a blind cast.

Reply:
It's a blind cast because the only way you can get there is if the selector is a `ModuleComponentSelector`, because, precisely, they are the only ones that have attributes. So it's not possible to have an error for anything else.

Reply:
Except that this method is called for _all_ selectors in case there is an error - See [from here](https://github.com/gradle/gradle/pull/5006/files/c7491d2642e7aa68231802d49fc314202dcf89df#diff-ec2b4b3ceac4be97bd7d2aa1944eb10fR249) to [there](https://github.com/gradle/gradle/pull/5006/files/c7491d2642e7aa68231802d49fc314202dcf89df#diff-ec2b4b3ceac4be97bd7d2aa1944eb10fR272) (Damn second link does not work well because section is folded by default - it is meant to go to line 272 (new numbering)

Reply:
Ok I see what you mean. I'm not sure it's even possible to have 2 different kinds of selectors for the same component id. The only case I can think of would be a project selector with a component selector, but I'm not sure this can ever happen in practice. I can add a safe guard, I'm not sure what we should do in that case. Probably show an empty attribute set?

Reply:
BTW I will not make the change as part of this PR, since the next one modifies this code to provide better error messages.